### PR TITLE
Image Feature Extraction pipeline

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Adds an ImageFeatureExtractor pipeline, which uses vision models for feature extraction. This is based off the `FeatureExtractionPipeline`, but is compatible with models which use image processors. 

```py
from transformers import pipeline

extractor = pipeline(model="google/vit-base-patch16-224", task="image-feature-extraction")
result = extractor("https://huggingface.co/datasets/Narsil/image_dummy/raw/main/parrots.png", return_tensors=True)
```

This was added instead of adapting the current feature-extractor pipeline because of the conditional handling required for the tokenizers versus image processors

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Review by osanseviero — COMMENTED** [review]
Very nice! :fire:

File: src/transformers/pipelines/image_feature_extraction.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    Learn more about the basics of using a pipeline in the [pipeline tutorial](../pipeline_tutorial)
+
+    This image feature extraction pipeline can currently be loaded from [`pipeline`] using the task identifier:
+    `"image-feature-extraction"`.
+
+    All vision models may be used for this pipeline. See a list of all models, including community-contributed models on
+    [huggingface.co/models](https://huggingface.co/models).
+
+    Arguments:
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py:35]: Yep! I've made a proposal PR here: https://github.com/huggingface/transformers/pull/28439

File: src/transformers/pipelines/image_feature_extraction.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            return model_outputs[0].tolist()
+        elif self.framework == "tf":
+            return model_outputs[0].numpy().tolist()
+
+    def __call__(self, *args, **kwargs):
+        """
+        Extract the features of the input(s).
+
+        Args:
+            args (`str` or `List[str]`): One or several texts (or one list of texts) to get the features of.
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py]: Done in d048cee

File: src/transformers/pipelines/image_feature_extraction.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            Model card attributed to the model for this pipeline.
+        framework (`str`, *optional*):
+            The framework to use, either `"pt"` for PyTorch or `"tf"` for TensorFlow. The specified framework must be
+            installed.
+
+            If no framework is specified, will default to the one currently installed. If no framework is specified and
+            both frameworks are installed, will default to the framework of the `model`, or to PyTorch if no model is
+            provided.
+        task (`str`, defaults to `""`, *optional*, defaults to `""`):
+            A task-identifier for the pipeline.
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py]: Yep - it's actually included as an input arg for pretty much all the pipelines, but I agree it doesn't make sense when the task is already specified. 

Removed in d048cee

File: tests/models/clip/test_modeling_clip.py [hunk 4]
```diff
@@ -477,7 +477,9 @@ def prepare_config_and_inputs_for_common(self):
 @require_torch
 class CLIPModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase):
     all_model_classes = (CLIPModel,) if is_torch_available() else ()
-    pipeline_model_mapping = {"feature-extraction": CLIPModel} if is_torch_available() else {}
+    pipeline_model_mapping = (
+        {"feature-extraction": CLIPModel, "image-feature-extraction": CLIPVisionModel} if is_torch_available() else {}
```
Comment (amyeroberts) [inline tests/models/clip/test_modeling_clip.py:481]: @ydshieh Is this how the pipeline mapping should be added for composite models? Or would it be better to add under `CLIPVisionModelTest` ?

**Comment by ydshieh** [comment]
We will need to make some changes in 

https://github.com/amyeroberts/transformers/blob/d048cee154ddb36d4eaf0a072f5ffabd9c0bef2d/tests/models/align/test_modeling_align.py#L43

to make the test of `"image-feature-extraction"` works.

File: tests/models/clip/test_modeling_clip.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/clip/test_modeling_clip.py:481]: The problem is that we don't have `hf-internal-testing/tiny-random-CLIPVisionModel`: the tiny model are only created for the top level model class not their components.

File: tests/pipelines/test_pipelines_image_feature_extraction.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+
+# We will verify our results on an image of cute cats
+def prepare_img():
+    image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
+    return image
+
+
+@is_pipeline_test
+class FeatureExtractionPipelineTests(unittest.TestCase):
```
Comment (ydshieh) [inline tests/pipelines/test_pipelines_image_feature_extraction.py:47]: Let's use `ImageFeatureExtractionPipelineTests` maybe

File: tests/models/clip/test_modeling_clip.py (hunk 4 above)
Comment (amyeroberts) [reply tests/models/clip/test_modeling_clip.py:481]: OK - I'll just remove for all the vision component models

File: tests/pipelines/test_pipelines_image_feature_extraction.py (hunk 5 above)
Comment (amyeroberts) [reply tests/pipelines/test_pipelines_image_feature_extraction.py]: Yep! Updated now

**Review by ydshieh — COMMENTED** [review]
Hi, sorry for a bit delay.

A first batch of reviews, mostly for the new file of image feature extractor implementation.

File: src/transformers/pipelines/image_feature_extraction.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            Device ordinal for CPU/GPU supports. Setting this to -1 will leverage CPU, a positive will run the model on
+            the associated CUDA device id.
+        torch_dtype (`str` or `torch.dtype`, *optional*):
+            Sent directly as `model_kwargs` (just a simpler shortcut) to use the available precision for this model
+            (`torch.float16`, `torch.bfloat16`, ... or `"auto"`).
+    """
+
+    def _sanitize_parameters(self, return_tensors=None, **kwargs):
+        preprocess_params = kwargs.pop("preprocess_kwargs", {})
+        postprocess_params = kwargs.pop("postprocess_kwargs", {})
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py]: I've updated in f346642 s.t.: 
* We don't accept `postprocess_kwargs` as an input
* We still create a `postprocess_params` which contains `return_tensors` to return
* We take `image_processor_kwargs` as input. This is used to create `preprocessor_params`, which `timeout` is added to. 

LMKWYT!

File: src/transformers/pipelines/image_feature_extraction.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    transformer, which can be used as features in downstream tasks.
+
+    Example:
+
+    ```python
+    >>> from transformers import pipeline
+
+    >>> extractor = pipeline(model="google/vit-base-patch16-224", task="image-feature-extraction")
+    >>> result = extractor("https://huggingface.co/datasets/Narsil/image_dummy/raw/main/parrots.png", return_tensors=True)
+    >>> result.shape  # This is a tensor of shape [1, sequence_lenth, hidden_dimension] representing the input string.
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py]: Done!

File: tests/pipelines/test_pipelines_image_feature_extraction.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    @require_tf
+    def test_return_tensors_tf(self):
+        feature_extractor = pipeline(
+            task="image-feature-extraction", model="hf-internal-testing/tiny-random-vit", framework="tf"
+        )
+        img = prepare_img()
+        outputs = feature_extractor(img, return_tensors=True)
+        self.assertTrue(tf.is_tensor(outputs))
+
+    def get_shape(self, input_, shape=None):
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/pipelines/test_pipelines_image_feature_extraction.py]: Removed :)

File: tests/pipelines/test_pipelines_image_feature_extraction.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    @require_torch
+    def test_image_processing_small_model_pt(self):
+        feature_extractor = pipeline(
+            task="image-feature-extraction", model="hf-internal-testing/tiny-random-vit", framework="pt"
+        )
+        img = prepare_img()
+        outputs = feature_extractor(img)
+        self.assertEqual(
+            nested_simplify(outputs[0][0]),
+            [-1.417, -0.392, -1.264, -1.196, 1.648, 0.885, 0.56, -0.606, -1.175, 0.823, 1.912, 0.081, -0.053, 1.119, -0.062, -1.757, -0.571, 0.075, 0.959, 0.118, 1.201, -0.672, -0.498, 0.364, 0.937, -1.623, 0.228, 0.19, 1.697, -1.115, 0.583, -0.981])  # fmt: skip
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/pipelines/test_pipelines_image_feature_extraction.py]: Removed :)

File: src/transformers/pipelines/image_feature_extraction.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+from .base import GenericTensor, Pipeline
+
+
+if is_vision_available():
+    from ..image_utils import load_image
+
+
+class ImageFeatureExtractionPipeline(Pipeline):
+    """
+    Image feature extraction pipeline using no model head. This pipeline extracts the hidden states from the base
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py]: Done here and for `FeatureExtractionPipeline's`

File: src/transformers/pipelines/image_feature_extraction.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+            Reference to the object in charge of parsing supplied pipeline parameters.
+        device (`int`, *optional*, defaults to -1):
+            Device ordinal for CPU/GPU supports. Setting this to -1 will leverage CPU, a positive will run the model on
+            the associated CUDA device id.
+        torch_dtype (`str` or `torch.dtype`, *optional*):
+            Sent directly as `model_kwargs` (just a simpler shortcut) to use the available precision for this model
+            (`torch.float16`, `torch.bfloat16`, ... or `"auto"`).
+    """
+
+    def _sanitize_parameters(self, return_tensors=None, image_processor_kwargs=None, **kwargs):
```
Comment (ydshieh) [inline src/transformers/pipelines/image_feature_extraction.py:61]: `image_processor_kwargs` could be before `return_tensors` (as it has more weight over `return_tensors`)

Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py]: Swapped ♻️

File: src/transformers/pipelines/image_feature_extraction.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        modelcard (`str` or [`ModelCard`], *optional*):
+            Model card attributed to the model for this pipeline.
+        framework (`str`, *optional*):
+            The framework to use, either `"pt"` for PyTorch or `"tf"` for TensorFlow. The specified framework must be
+            installed.
+
+            If no framework is specified, will default to the one currently installed. If no framework is specified and
+            both frameworks are installed, will default to the framework of the `model`, or to PyTorch if no model is
+            provided.
+        args_parser ([`~pipelines.ArgumentHandler`], *optional*):
```
Comment (ydshieh) [inline src/transformers/pipelines/image_feature_extraction.py:51]: Might be better to have `image_processor_kwargs` in the doc.

(and we probably need to check the rendered doc page to see if all the arguments defined in the parent class are included - no need to do this in this PR though if anything wrong)

File: tests/models/vilt/test_modeling_vilt.py [hunk 13]
```diff
@@ -227,11 +227,7 @@ class ViltModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase):
         if is_torch_available()
         else ()
     )
-    pipeline_model_mapping = (
-        {"feature-extraction": ViltModel, "visual-question-answering": ViltForQuestionAnswering}
-        if is_torch_available()
-        else {}
-    )
+    pipeline_model_mapping = {"visual-question-answering": ViltForQuestionAnswering} if is_torch_available() else {}
```
Comment (ydshieh) [inline tests/models/vilt/test_modeling_vilt.py:230]: nit: just wondering why `"feature-extraction": ViltModel` is not kept (like in other places)

File: tests/models/git/test_modeling_git.py [hunk 14]
```diff
@@ -399,7 +399,11 @@ class GitModelTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMixin,
     all_model_classes = (GitModel, GitForCausalLM) if is_torch_available() else ()
     all_generative_model_classes = (GitForCausalLM,) if is_torch_available() else ()
     pipeline_model_mapping = (
-        {"feature-extraction": GitModel, "image-to-text": GitForCausalLM, "text-generation": GitForCausalLM}
+        {
+            "feature-extraction": GitModel,
+            "image-to-text": GitForCausalLM,
+            "text-generation": GitForCausalLM,
+        }
```
Comment (ydshieh) [inline tests/models/git/test_modeling_git.py:406]: nit: wondering if you manually change this. It doesn't look our styling tool change the previous format.

**Review by ydshieh — APPROVED** [review]
Thank you a lot. LGTM.

Just 3 nits / questions.

Remark (just FYI):  the attribute `pipeline_model_mapping` is generated via a script. And usually we don't modify them manually. It's fine here in this PR. But next time when I need to run the script again, it will add back `"feature-extraction"` anyway (despite the corresponding test will be skip).

**Comment by amyeroberts** [comment]
> Remark (just FYI): the attribute pipeline_model_mapping is generated via a script. And usually we don't modify them manually. It's fine here in this PR. But next time when I need to run the script again, it will add back "feature-extraction" anyway (despite the corresponding test will be skip).

@ydshieh OK, thanks for flagging. Is there anything else I need to do to make sure that `"image-feature-extraction"` will be automatically added / not deleted when running this script?

File: tests/models/git/test_modeling_git.py (hunk 14 above)
Comment (amyeroberts) [reply tests/models/git/test_modeling_git.py]: I think it was probably a trailing comma on my part. I've updated to be one line again

**Comment by ydshieh** [comment]
> > Remark (just FYI): the attribute pipeline_model_mapping is generated via a script. And usually we don't modify them manually. It's fine here in this PR. But next time when I need to run the script again, it will add back "feature-extraction" anyway (despite the corresponding test will be skip).
> 
> @ydshieh OK, thanks for flagging. Is there anything else I need to do to make sure that `"image-feature-extraction"` will be automatically added / not deleted when running this script?

No, no action required from your side :-) in this PR

File: src/transformers/pipelines/image_feature_extraction.py (hunk 12 above)
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py]: Added!

File: tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py [hunk 15]
```diff
@@ -323,6 +324,10 @@ def test_encoder_decoder_model_output_attentions(self):
         input_ids_dict = self.prepare_config_and_inputs()
         self.check_encoder_decoder_model_output_attentions(**input_ids_dict)
 
+    # FIXME @ydshieh: flaky test
+    @is_flaky(
+        description="Fails on distributed runs e.g.: https://app.circleci.com/pipelines/github/huggingface/transformers/83611/workflows/666b01c9-1be8-4daa-b85d-189e670fc168/jobs/1078635/tests#failed-test-0"
+    )
```
Comment (ydshieh) [inline tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py:330]: I will

File: tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py [hunk 16]
```diff
@@ -323,6 +324,10 @@ def test_encoder_decoder_model_output_attentions(self):
         input_ids_dict = self.prepare_config_and_inputs()
         self.check_encoder_decoder_model_output_attentions(**input_ids_dict)
 
+    # FIXME @ydshieh: flaky test
```
Comment (amyeroberts) [inline tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py:327]:
```suggestion
    # FIXME @gante: flaky test
```

File: src/transformers/pipelines/image_feature_extraction.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
+from typing import Dict
```
Comment (ArthurZucker) [inline src/transformers/pipelines/image_feature_extraction.py:1]: we need a licence

File: tests/pipelines/test_pipelines_image_feature_extraction.py [hunk 18]
```diff
@@ -0,0 +1,157 @@
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
Comment (ArthurZucker) [inline tests/pipelines/test_pipelines_image_feature_extraction.py:1]:
```suggestion
# Copyright 2024 The HuggingFace Team. All rights reserved.
```

**Review by ArthurZucker — APPROVED** [review]
LGTM 😉

**Comment by ArthurZucker** [comment]
FYI @Rocketknight1

File: src/transformers/pipelines/image_feature_extraction.py (hunk 17 above)
Comment (amyeroberts) [reply src/transformers/pipelines/image_feature_extraction.py:1]: I'll add one 👍  The reason there isn't one here is that there isn't one for all of the `pipelines/xxx.py` files.

**Pull request merged**